Apps/Src/SendCarCAN.c

```diff
@@ -1,35 +1,89 @@
-#include "Tasks.h"
-#include "CANbus.h"
-#include "CAN_Queue.h"
 #include "common.h"
+#include "os_cfg_app.h"
+#include "CANbus.h"
+#include "Minions.h"
+#include "Contactors.h"
+#include "Pedals.h"
+#include "Tasks.h"
+#include "SendCarCAN.h"
+
+// fifo
+#define FIFO_TYPE CANDATA_t
+#define FIFO_SIZE 256
+#define FIFO_NAME SendCarCAN_Q
+#include "fifo_protected.h"
+
+static SendCarCAN_Q_t CANFifo;
+
+#define IO_STATE_TMR_DLY_MS 250u
+#define IO_STATE_TMR_DLY_TS ((IO_STATE_TMR_DLY_MS * OS_CFG_TMR_TASK_RATE_HZ) / (1000u))
+static OS_TMR IOStateTmr;
+
+static void putIOState(void *p_tmr, void *p_arg);
 
 /**
- * @file SendCarCAN.c
- * @brief Implements the SendCarCAN Task
- * 
- * Resends the feedback from the motor controller
- * 
- */
+ * @brief Initialize SendCarCAN
+*/
+void SendCarCAN_Init(){
+    OS_ERR err;
+    
+    SendCarCAN_Q_renew(&CANFifo);
+    
+    // Set up timer to put car data message every IO_STATE_TMR_DLY_MS ms
+    OSTmrCreate(&IOStateTmr,
+                "IO State CAN Message Timer",
+                0,
+                IO_STATE_TMR_DLY_TS,
+                OS_OPT_TMR_PERIODIC,
+                putIOState,
+                NULL,
+                &err
+                );
+}
 
 /**
- * @brief Sends the feedback from the motor controller to be read by telemetry and 
- * sends pedal, switch, light, and contactor information to be read by telemetry
- * 
- * @param p_arg 
- */
+ * @brief Wrapper to put new message in the CAN queue
+*/
+void SendCarCAN_Put(CANDATA_t message){
+    SendCarCAN_Q_put(&CANFifo, message);
+}
+
+/**
+ * @brief Grabs the latest messages from the queue and sends over CarCAN
+*/
 void Task_SendCarCAN(void *p_arg){
-    CANDATA_t motorMsg;
-    OS_ERR err;
+    CANDATA_t message;
 
     while (1) {
-        // Send motor msg
-        CAN_Queue_Pend(&motorMsg);
-        CANbus_Send(motorMsg, true, CARCAN);
-
-        // Delay of few milliseconds (100)
-        OSTimeDlyHMSM(0, 0, 0, 100, OS_OPT_TIME_HMSM_STRICT, &err);
-        if (err != OS_ERR_NONE){
-            assertOSError(OS_UPDATE_VEL_LOC, err);
-        }
+        SendCarCAN_Q_get(&CANFifo, &message);
+        CANbus_Send(message, true, CARCAN);
+        
```

Comment:
Do we want a delay here? I'm guessing it's fine if we always try to send when something gets in the FIFO

Reply:
I figured we don't need a delay, since SendCarCAN_Q does a semaphore pend which will act as a scheduling point anyways.